### PR TITLE
Feat/instruct

### PR DESCRIPTION
**Pull request by shamuiscoding**
@engineerA314 Need fix evaluation (to be able to return recall and precision) before merging

### CODE REVIEW

**Comment by engineerA314** [comment]
still debugging for `MIPROInstruct`

File: ape/proposer/instruct_by_score.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            json.dumps(message) for message in base_prompt.messages
+        ]
+        base_prompt_messages_str = "\n".join(base_prompt_messages)
+
+        try:
+            # Generate new instruction from Prompt.from_filename("gen-instruction-with-eval")
+            # input : base_prompt, evaluation_result, evaluation_function, tip, response_format
+            new_instruction_text = await self.generate_instructions(
+                base_prompt=base_prompt_messages_str,
+                evaluation_result=str(evaluation_result),
```
Comment (engineerA314) [inline ape/proposer/instruct_by_score.py]: what value should be sent to LLM as `evaluation_result`?
`List[EvaluationResult]`, or some part of it, or final score(float), or everything, or analysis of intermediate values?
@shamuiscoding

File: ape/ape_prompts/gen-instruction-with-eval.prompt [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+A good way to structure the prompt is to add format/instructions to the system prompt, and user inputs to the user prompt.
+
+If you would like to use few shot examples, use the reserved variable {{_FEWSHOT_}} - just insert it into the prompt and a list of task demonstrations will be filled in afterwards. Your goal is to generate the proper and complete instructions that can be used without revision.
+
+> Use brackets {{}} ONLY for input variables and the reserved few shot variable. Use [] instead for placeholders.
+
+> If the response format is specified as json_object, ensure the instruction includes the word "json".
+
+> Do not use backticks except for the CDATA section.
```
Comment (aschung01) [inline ape/ape_prompts/gen-instruction-with-eval.prompt]: Let's remove this

File: ape/proposer/instruct_by_score.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+TIPS = {
+    "none": "Make it better",
+    "one_side_first": "Try to optimize a certain part of the evaluation score first. For example, if the evaluation is based on both precision and recall, try to optimize for recall first. Be extreme at times.",
+    "double_down": "Find what the prompt is already doing well and double down on it. For example, if the prompt already does a good job with precision, try to improve that even further with various strategies. Be extreme at times",
+    "weakness_first": "Find what the prompt is already doing poorly based on the evaluation score and metric and try to improve that. For example, if the prompt has low recall, try to improve that with various strategies. Be extreme at times",
+    "think": "Try to think about what the evaluation score and metric are actually measuring. Then, try to optimize for that. For example, if the evaluation score is calculated based on the average length of the generated instructions, try to optimize for that. Be extreme at times",
+}
+
+
+class InstructByScore(Proposer):
```
Comment (aschung01) [inline ape/proposer/instruct_by_score.py]: Rename this to `EvaluationDrivenProposer`?

**Review by aschung01 — CHANGES_REQUESTED** [review]

**Review by aschung01 — APPROVED** [review]

**Pull request merged**